src/org/etools/j1939_84/controllers/part05/Part05Step02Controller.java

```diff
@@ -52,11 +58,45 @@
 
     @Override
     protected void run() throws Throwable {
+
+        clearPreviousDM12Packets();
+
         // 6.5.2.1.a Global DM12 [(send Request (PGN 59904) for PGN 65236 (SPNs 1213-1215, 1706, and 3038)]).
+        var globalPackets = getDiagnosticMessageModule().requestDM12(getListener()).getPackets();
+
         // 6.5.2.2.a Fail if no OBD ECU reporting MIL on. See Section A.8 for allowed values.
+        boolean noMilOn = globalPackets.stream().noneMatch(p -> (p.getMalfunctionIndicatorLampStatus() == ON));
+        if (noMilOn) {
+            addFailure("6.5.2.2.a - No OBD ECU reported MIL on");
+        }
+
         // 6.5.2.2.b Fail if all OBD ECUs report no DM12 DTC set.
-        // 6.5.2.2.c Fail if DM12 DTC reported does not match the DM6 DTC SPN and FMI reported from step 6.4.2.
-        // 6.5.2.2.d Fail if any ECU reporting MIL as ON, flashing. See Section A.8 for allowed values.
+        boolean noDM12 = globalPackets.stream().noneMatch(p -> (p.getDtcs().size() > 0));
+        if (noDM12) {
+            addFailure("6.5.2.2.b - All OBD ECUs report no DM12 DTCs");
+        }
+
+        globalPackets.forEach(packet -> {
```

Comment:
nit: I like to cycle through the packets once for each failure just to keep the code cleaner.